src/vm/jitinterface.cpp

```diff
@@ -6026,14 +6026,28 @@ CorInfoHelpFunc CEEInfo::getNewHelper(CORINFO_RESOLVED_TOKEN * pResolvedToken, C
 }
 
 /***********************************************************************/
-CorInfoHelpFunc CEEInfo::getNewHelperStatic(MethodTable * pMT)
+CorInfoHelpFunc CEEInfo::getNewHelperStatic(MethodTable * pMT, bool * pHasSideEffects)
 {
     STANDARD_VM_CONTRACT;
 
 
     // Slow helper is the default
     CorInfoHelpFunc helper = CORINFO_HELP_NEWFAST;
+    bool hasFinalizer = pMT->HasFinalizer();
 
+    if (pHasSideEffects != nullptr)
+    {
+#ifdef FEATURE_READYTORUN_COMPILER
+        if (IsReadyToRunCompilation())
+        {
+            *pHasSideEffects = hasFinalizer || !pMT->IsInheritanceChainFixedInCurrentVersionBubble();
+        }
+        else
+#endif
+        {
+            *pHasSideEffects = hasFinalizer;
+        }
+    }
 
     if (pMT->IsComObjectType())
     {
```

Comment:
Should the ComObjectType case have unconditional `*pHasSideEffects = true` ?

Reply:
I'm not sure. What is the side effect of allocating a ComObjectType? Why can't the jit remove a dead call to a helper allocating ComObjectType?

Reply:
Allocating COM objects has number of side-effects: It will initialize COM on the given thread, it will check validity of the COM object, .... .

Reply:
Makes sense, AllocateComObject_ForManaged is called when allocating COM objects. I updated the change, PTAL.